Examples: Add glTF compression example

zeux opened the pull request:
ToyCar.glb is exported from glTF-Sample-Models using gltfpack with
geometry compression via EXT_meshopt_compression and texture compression
via KHR_texture_basisu (using UASTC+zstd since ETC1S creates very
visible reflection artefacts).

The resulting .glb file is 1.9 MB, compared to original 6.9 MB; in the
resulting file, geometry takes 0.6 MB and textures take 1.3 MB; this is
compared to the original 3.5 MB of geometry and 3.4 MB of textures.

Fixes #21111

zeux committed:
Add glTF compression example

coffeemat.glb is exported from the source Sketchfab scene using gltfpack
with geometry compression via EXT_meshopt_compression and texture compression
via KHR_texture_basisu (using ETC1S/BasisLZ).

The source scene has 12 MB of geometry and 22 MB of texture data.
The compressed scene has 2 MB of geometry and 2.6 MB of texture data.

zeux commented:
This is how the scene looks like:

![image](https://user-images.githubusercontent.com/1106629/105123889-c3650c00-5a8d-11eb-834a-b1bbddf926a8.png)

donmccurdy commented on examples/webgl_loader_gltf_compressed.html:
```diff
@@ -0,0 +1,143 @@
+								if ( child.isMesh ) {
+
+									// TOFIX RoughnessMipmapper seems to be broken with WebGL 2.0
+									// roughnessMipmapper.generateMipmaps( child.material );
```
Probably best to omit RoughnessMipmapper, I'm not sure when this will be fixed.

zeux replied:
Good call especially since even if it's fixed it's not going to be compatible with compressed textures anyway! Fixed.

donmccurdy commented on examples/webgl_loader_gltf_compressed.html:
```diff
@@ -0,0 +1,143 @@
+						// gltfpack -i ~/glTF-Sample-Models/2.0/ToyCar/glTF/ToyCar.gltf -o ToyCar.glb -cc -tc -tu -tq 1
+						// The resulting model uses EXT_meshopt_compression (for geometry) and KHR_texture_basisu (for texture compression using UASTC+zstd)
+
+						loader.load( 'ToyCar.glb', function ( gltf ) {
```
One downside of this model is that it highlights our lack of support for `KHR_materials_transmission`, and `KHR_materials_sheen`. See #21000 and #16977. 😐 Is the front windshield not transparent at all without `KHR_materials_transmission`, or is that just glare in the screenshot?

zeux replied:
The windshield is indeed completely opaque :) And yeah I noticed the lack of sheen on cloth when comparing with the Babylon rendering...

zeux replied:
FWIW the nice thing about that model and the reason why I picked it is that it has a fair amount of geometry and texture data, and a host of different texture types - so it's a great example that captures the variety of data types. It doesn't have animation or morph targets but it's hard to pick a good all-in-one example...

donmccurdy replied:
Perhaps this one instead? https://sketchfab.com/3d-models/pony-cartoon-885d9f60b3a9429bb4077cfac5653cf9

zeux replied:
This one is very low-poly unfortunately :( I can try to find something else on Sketchfab I suppose although I'm not sure what the licensing restrictions are.

donmccurdy replied:
I do agree it's otherwise a nice example.

donmccurdy replied:
We're happy to include attribution on these examples, and non-commercial use restrictions should be fine for three.js as well. Or if you'd rather wait, perhaps someone will eventually (a) chime in with a better idea, or (b) implement transmission and sheen. 😁

zeux replied:
Less diverse texture set but how about https://sketchfab.com/3d-models/coffeemat-7fb196a40a6e4697aad9ca2f75c8b33d ?

donmccurdy replied:
Looks great to me!

donmccurdy reviewed (commented):
Thank you! Two comments, otherwise looks good to me.

zeux commented:
Replaced the scene with https://sketchfab.com/3d-models/coffeemat-7fb196a40a6e4697aad9ca2f75c8b33d:

![image](https://user-images.githubusercontent.com/1106629/105130893-b3a0f400-5a9c-11eb-911b-f13cafa9aba5.png)

The source scene has 12 MB of geometry and 22 MB of texture data.
The compressed scene has 2 MB of geometry and 2.6 MB of texture data (using ETC1S).

I've also squashed the commits to make sure the old .glb doesn't pollute the history.

mrdoob commented:
Thanks!

mrdoob commented:
I have simplified the example a bit: https://github.com/mrdoob/three.js/commit/067cf15b815150f24b0ba072fa688b40d37dc6f5

http://raw.githack.com/mrdoob/three.js/dev/examples/webgl_loader_gltf_compressed.html
